docs/adr/20220314-caching-build-tasks.md

```diff
@@ -0,0 +1,86 @@
+# Build Task Caching
+
+Date: 2022-03-14
+
+## Status
+
+Accepted
+
+## Context
+
+For a repo with a single build, the pipeline build will only be triggered if any changes trigger a new pipeline run.
+
+However if one has multiple builds in one repo, the pipeline build runs all build tasks even for those areas where nothing has changed.
+
+A simple scenario is a repo with a backend and frontend component whose build outputs are combined by a subsequent docker build in a subsequent deploy task.
+
+There is no good reason when only the frontend code changes that the backend build is redone as well and vice versa of course.
+
+## Decision
+
+Build task caching persists a build's produced files so that future builds can be skipped if the working directory specified in the build task did not change.
+
+The build files (output and reports) are cached in a dedicated cash area named `build-task` on the caching PVC described in the [caching ADR](20220225-caching.md).
+
+By default build task caching is not enabled for the following reasons:
+
+- Build task skipping has limited benefits for single build repos which is the main design focus of `ods-pipeline`. The build task cache could only be reused when the working dir did not change while some other change in the repo triggers a build. This might be the case when the working-dir is not at the root or when there is a merge commit without file changes.
```

Comment:
Given my above comment, do you think this still holds true? Or should we enable by default?

Reply:
One reason for keeping it opt-in is that there is also the fine print for ensuring the build is not using files outside their working directory.  This case is more likely in multi build repos and so this would speak for enabling it perhaps by default in single build or when working-dir is `.` only by default.
The overhead can get somewhat larger with lots of node_modules files. For some reason this appears not to be an issue in the pipeline where I did not see the copy with hard linking needing more than 2s . However on my Mac locally this sometimes tooks in the 20s range. 
All in all I don't feel that strongly about to not enable it. 
I could also imagine to have it off by default for the initial release with the intend of turning this on in the next release.